packages/arb-bridge-eth/contracts/arch/OneStepProofCommon.sol

```diff
@@ -312,42 +310,53 @@ abstract contract OneStepProofCommon {
             function(AssertionContext memory) internal view impl
         ) = opInfo(context.opcode);
 
-        // Update end machine gas remaining before running opcode
-        if (deductGas(context, gasCost)) {
-            return;
-        }
+        // Require the prover to submit the minimal number of stack items
+        require(
+            ((dataPopCount > 0 || !context.hadImmediate) && context.stack.length <= dataPopCount) ||
+                (context.hadImmediate && dataPopCount == 0 && context.stack.length == 1),
+            STACK_MANY
+        );
+        require(context.auxstack.length <= auxPopCount, AUX_MANY);
 
+        // Update end machine gas remaining before running opcode
         if (context.stack.length < dataPopCount) {
             // If we have insufficient values, reject the proof unless the stack has been fully exhausted
             require(
                 context.afterMachine.dataStack.hash() == Value.newEmptyTuple().hash(),
                 STACK_MISSING
             );
+            deductGas(context, ERROR_GAS_COST);
             // If the stack is empty, the instruction underflowed so we have hit an error
             handleError(context);
-            return;
-        }
-
-        if (context.auxstack.length < auxPopCount) {
+        } else if (context.auxstack.length < auxPopCount) {
             // If we have insufficient values, reject the proof unless the auxstack has been fully exhausted
             require(
                 context.afterMachine.auxStack.hash() == Value.newEmptyTuple().hash(),
                 AUX_MISSING
             );
+            deductGas(context, ERROR_GAS_COST);
             // If the auxstack is empty, the instruction underflowed so we have hit an error
             handleError(context);
-            return;
+        } else if (deductGas(context, gasCost)) {
+            handleError(context);
+        } else {
+            impl(context);
         }
 
-        // Require the prover to submit the minimal number of stack items
-        require(
-            ((dataPopCount > 0 || !context.hadImmediate) && context.stack.length == dataPopCount) ||
-                (context.hadImmediate && dataPopCount == 0 && context.stack.length == 1),
-            STACK_MANY
-        );
-        require(context.auxstack.length == auxPopCount, AUX_MANY);
-
-        impl(context);
+        if (context.errorOccurred) {
+            if (context.afterMachine.errHandlerHash == CODE_POINT_ERROR) {
+                context.afterMachine.setErrorStop();
+            } else {
+                // Clear error
+                context.errorOccurred = false;
+                context.afterMachine.instructionStackHash = context.afterMachine.errHandlerHash;
+
+                if (!(context.hadImmediate && dataPopCount == 0)) {
```

Comment:
Don't we need to do something with the stack when the condition in this `if` is false?

Reply:
What this should do is clear the stack unless the instruction had an immediate and didn't have any arguments from the spec, in which case by the current behavior, the immediate value should stay on the stack. The semantics of this are the same as though the immediate is pushed to the stack before the instruction is run